proposals/pr002-hunkspace.md

```diff
@@ -0,0 +1,545 @@
+independent piece of software.  A localization clearinghouse would allow
+the OS to centralize localization decisions.
+
+A way to resolve is to devise some form of a mini "rendering protocol"
+where the screen reader provides rendering hints, and the voice UI
+renderer follows the hints.  The claim here is that this specific
+pattern is common enough that localization could benefit from a common
+specification of that rendering protocol.
+
+# Detailed Proposal
+
+Allow independent third parties to put new non-overlapping points into
+a common hunkspace, such that for each platform a single data provider
+can be used to provide any localized resource.
+
+Specify a mechanism by which this can be done, e.g. by adopting an
+URI-like encoding of hunkspace points, and providing a two-way mapping
+between an URI and an "unpacked" version of a hunkspace key.
+
+**Why this complication?** It allows third party contributions to localized
+software in trusted computing environments.  For example, third party provision
+of localized resources, as described in the Fuchsia use case.
+
+**Why should ICU4X care?** Strictly speaking perhaps it doesn't need to. One
+can build a "hunkspace" data provider given a cldr data provider, and fixing
+the "source" dimension to be "org.unicode.cldr" (see below).  An advantage of a
+commonly understood specification is that it may motivate software authors to
+provide mutually compatible data provider interfaces.
+
+**Do we even need common schemas like this?** In a world where localizations
+are islands of functionality, maybe the answer is no.  A CLDR data provider can
+exist on its own.  Firefox has its own elaborate L10N framework.  Fuchsia OS
+has use cases where platform-level I18N/L10N data provision is needed, but
+these use cases can live inside Fuchsia indefinitely.  It starts to matter if
+these worlds get to interact at some level. For example Firefox L10N can not
+exist in its current extensible form on Fuchsia, due to the tight controls that
+the OS puts in place on the runtime. (You'd have to download a prebuilt version
+for your locale, from Mozilla.)  Having a cross-platform data provider
+specification could make that a non-issue.]
+
+## Unique IDs
+
+Every localizable resource should get a unique identifier.  This allows
+a platform-level localization clearinghouse (from the example above) to
+be a one-stop data provider for localization needs.
+
+For example, the [current proposal of the icu4x data provider for
+CLDR][dpcldr] uses a 4-dimensional key space consisting of:
+
+- A Language identifier;
+- a Category;
+- a Key;
+- and a Payload.
+
+[dpcldr]: https://github.com/unicode-org/icu4x/blob/master/components/data-provider/src/lib.rs#L52
+
+This key space is intended to fit any CLDR resource according to the data
+mapping which is internally defined by CLDR rules.
+
+It does not represent a fictitious resource string "Hello, {user}" from
```

Comment:
nit: formatting: change to code regarding the resource string:  (don't need to do this for every quotation mark set, but just for anything meant to be seen as code)

"Hello, {user}" vs `"Hello, {user}"`